File: lab05/matlab/Problem2.m

N_values = [2, 4, 6, 8];
omega_3dB = 2*pi*100; 
w = 0:0.1:2000;
ang = zeros(4,length(w));
Hdec= zeros(4,length(w));
Hlin = zeros(4,length(w));

i=1;
for N = N_values
    poles = omega_3dB*exp(1i*(pi/2 + 1/2 * pi/N + ((1:N)-1)*pi/N));
    wzm = prod(-poles);
    a = poly(poles);
    b = wzm;
    h = freqs(b,a,w);
    
    ang(i,:) = angle(h);
    Hdec(i,:)= 20*log10(abs(h));
    Hlin(i,:) = abs(h);

    i=i+1;
    
end
figure;
hold on;
grid on;
wlog = logspace(0,2,20001)
for row=1:4
semilogx(w./(2*pi),Hdec(row,:));
end
legend("2","4","6","8");
title("Charakterystyka A-cz skala logarytmiczna")

figure;
hold on;
for row=1:4
plot(w./(2*pi),Hlin(row,:));
end
legend("2","4","6","8");
title("Charakterystyka A-cz skala liniowa")

figure;
hold on;
for row=1:4
plot(w./(2*pi),ang(row,:));
end
legend("2","4","6","8");
title("Charakterystyka cz-f");

figure;
poles4 = omega_3dB*exp(1i*(pi/2 + 1/2 * pi/4 + ((1:4)-1)*pi/4));
a = poly(poles4);
b = wzm;
H = tf(b,a);
[y,tOut] =impulse(H);
plot(tOut,y );
title("Odpowiedź impulsowa")
figure;
[y,tOut] = step(H);
plot(tOut,y)
title("Odpowiedź na skok jednostkowy");

File: lab05/problem2.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.signal import freqs, TransferFunction, impulse, step

N_values = [2, 4, 6, 8]
omega_3dB = 2 * np.pi * 100
w = np.arange(0, 2000, 0.1)
ang = np.zeros((4, len(w)))
H_dec = np.zeros((4, len(w)))
H_lin = np.zeros((4, len(w)))


for i, N in enumerate(N_values):
    poles = omega_3dB * \
        np.exp(1j * (np.pi / 2 + 1 / 2 * np.pi / N + (np.arange(N) * np.pi / N)))
    amp = np.prod(-poles)
    a = np.poly(poles)
    b = amp
    H = TransferFunction(b, a)
    w_out, h = freqs(H.num, H.den, w)

    ang[i, :] = np.angle(h)
    H_dec[i, :] = 20 * np.log10(np.abs(h))
    H_lin[i, :] = np.abs(h)

plt.figure()
plt.subplot(2, 1, 1)
plt.grid(True)
for row in range(4):
    plt.plot(w_out / (2 * np.pi), H_dec[row, :])
plt.legend(N_values)
plt.title("Charakterystyka Amplitudy w Skali Logarytmicznej")
plt.ylabel("Amplituda [dB]")

plt.subplot(2, 1, 2)
plt.grid(True)
for row in range(4):
    plt.plot(w_out / (2 * np.pi), H_lin[row, :])
plt.legend(N_values)
plt.title("Charakterystyka Amplitudy w Skali Liniowej")
plt.xlabel("Częstotliwość [Hz]")
plt.ylabel("Amplituda")

plt.figure()
plt.grid(True)
for row in range(4):
    plt.plot(w_out / (2 * np.pi), ang[row, :])
plt.legend(N_values)
plt.title("Charakterystyka Fazowa")
plt.xlabel("Częstotliwość [Hz]")
plt.ylabel("Kąt [rad]")

plt.figure()
plt.grid(True)
for row in range(4):
    plt.plot(w_out / (2 * np.pi), np.unwrap(ang[row, :]))
plt.legend(N_values)
plt.title("Charakterystyka Fazowa [unwraped]")
plt.xlabel("Częstotliwość [Hz]")
plt.ylabel("Kąt [rad]")



plt.figure()
plt.grid(True)


poles4 = omega_3dB * \
    np.exp(1j * (np.pi / 2 + 1 / 2 * np.pi / 4 + (np.arange(4) * np.pi / 4)))
a = np.poly(poles4)
b = amp
H = TransferFunction(b, a)
tOut, y = impulse(H)
plt.plot(tOut, y)
plt.title("Odpowiedź Impulsowa")
plt.xlabel("Czas [s]")
plt.ylabel("Amplituda")

plt.figure()
plt.grid(True)
tOut, y = step(H)
plt.plot(tOut, y)
plt.title("Odpowiedź na Skok Jednostkowy")
plt.xlabel("Czas [s]")
plt.ylabel("Amplituda")

plt.show()
